cl/lib/elasticsearch_utils.py

```diff
@@ -69,18 +112,26 @@ def build_fulltext_query(fields: list[str], value: str) -> QueryString | List:
                 value = value.replace(
                     f"docketNumber:{match}", f"docketNumber:{replacement}"
                 )
+
         q_should = [
             Q(
-                "multi_match",
-                query=value,
+                "query_string",
                 fields=fields,
-                type="phrase",
-                operator="AND",
+                query=value,
+                quote_field_suffix=".exact",
+                default_operator="AND",
                 tie_breaker=0.3,
             ),
-            Q("query_string", query=value, default_operator="AND"),
+            Q(
+                "query_string",
+                query=value,
+                quote_field_suffix=".exact",
+                default_operator="AND",
+                type="phrase",
+            ),
```

Comment:
Indeed, I'm not sure if having two `query_strings` makes the query twice as expensive (or if elasticsearch makes a kind of optimization internally) so far this was the approach I've found to be able to perform advanced queries and phrase queries in a single text input (like in Solr). I could do a quick benchmarking to measure its impact vs a single `query_string` and also report the effects of only having one `query_string`. This way, we can make an informed decision or explore alternative solutions.

Reply:
That'd be great, thanks. I expect we'll have more benchmarking questions, so it's probably worth the time.

Reply:
@mlissner I performed a benchmark using [rally](https://esrally.readthedocs.io/en/stable/) which is the official Elasticsearch benchmarking tool.

Here are the results:

**search_1**
Using the query approach present in this PR, two query strings, one with the type `best_fields` (default type) and limiting the fields to look at. And the second type `phrase` that looks in all the fields.

```
"query":{
       "bool":{
          "should":[
             {
                "query_string":{
                   "fields":[
                      "court",
                      "court_citation_string",
                      "judge",
                      "text^1.0",
                      "caseName^4.0",
                      "docketNumber^2.0"
                   ],
                   "query":"*",
                   "quote_field_suffix":".exact",
                   "default_operator":"AND",
                   "tie_breaker":0.3
                }
             },
             {
                "query_string":{
                   "query":"*",
                   "quote_field_suffix":".exact",
                   "default_operator":"AND",
                   "type":"phrase"
                }
             }
          ]
       }
    }
```

The results related to the query are:

```
|                                                 Min Throughput | search_1 | 7782.24        |  ops/s |
|                                                Mean Throughput | search_1 | 7782.24        |  ops/s |
|                                              Median Throughput | search_1 | 7782.24        |  ops/s |
|                                                 Max Throughput | search_1 | 7782.24        |  ops/s |
```

These are the queries performed by second.
Full results here:
[search_1.txt](https://github.com/freelawproject/courtlistener/files/11666851/search_1.txt)


**search_2**
This is the query approach I'm using in the RT Alerts project where it was needed to limit the fields in the second query, so we'll have two query strings, one with the type `best_fields` (default type) and limiting the fields to look at. And the second type `phrase` and also limits the fields to look at.

```
{
   "query":{
      "bool":{
         "should":[
            {
               "query_string":{
                  "fields":[
                     "court",
                     "court_citation_string",
                     "judge",
                     "text^1.0",
                     "caseName^4.0",
                     "docketNumber^2.0"
                  ],
                  "query":"*",
                  "quote_field_suffix":".exact",
                  "default_operator":"AND",
                  "tie_breaker":0.3
               }
            },
            {
               "query_string":{
                  "fields":[
                     "court",
                     "court_citation_string",
                     "judge",
                     "text^1.0",
                     "caseName^4.0",
                     "docketNumber^2.0"
                  ],
                  "query":"*",
                  "quote_field_suffix":".exact",
                  "default_operator":"AND",
                  "type":"phrase"
               }
            }
         ]
      }
   }
```
The results related to the query are:

```
|                                                 Min Throughput | search_2 | 7847.24        |  ops/s |
|                                                Mean Throughput | search_2 | 7847.24        |  ops/s |
|                                              Median Throughput | search_2 | 7847.24        |  ops/s |
|                                                 Max Throughput | search_2 | 7847.24        |  ops/s |
```
Full results here:
[search_2.txt](https://github.com/freelawproject/courtlistener/files/11666852/search_2.txt)


**search_3**
This approach is using only 1 query string, however, this approach has some issues that I'll explain later.

```
"query":{
      "query_string":{
         "fields":[
            "court",
            "court_citation_string",
            "judge",
            "text^1.0",
            "caseName^4.0",
            "docketNumber^2.0"
         ],
         "query":"*",
         "quote_field_suffix":".exact",
         "default_operator":"AND",
         "tie_breaker":0.3
      }
   }
```
The results related to the query are:

```
|                                                 Min Throughput | search_2 | 7847.24        |  ops/s |
|                                                Mean Throughput | search_2 | 7847.24        |  ops/s |
|                                              Median Throughput | search_2 | 7847.24        |  ops/s |
|                                                 Max Throughput | search_2 | 7847.24        |  ops/s |
```
Full results here:
[search_3.txt](https://github.com/freelawproject/courtlistener/files/11666853/search_3.txt)


In brief, simpler queries lead to better performance.

`search_1` two queries, one of them looking in all fields: 7782.24  ops/s
`search_2` two queries, both of them limiting the fields to look at: 7847.24 ops/s
`search_3` one query limiting the fields to look at: 7961.26 ops/s

However, the performance impact is not proportional to the number of queries of fields to look at.


Here are the issues I found using one `query_string` approach:
Using one `query_string` (best_fields):

The results it returns doesn't consider the order of words to assign scores, leading to showing results sorted without considering phrase order.

e.g in this example the query is `Lynch Loretta` so results with `Lynch` appearing before `Loretta` should be first, and here is not the case:
![Screenshot 2023-06-05 at 23 59 53](https://github.com/freelawproject/courtlistener/assets/486004/00d8e0b9-4415-4250-9c13-a043084b9f15)

Also in this approach highlighting is not working as expected, here `deposition` should be highlighted too.
![Screenshot 2023-06-05 at 23 59 14](https://github.com/freelawproject/courtlistener/assets/486004/b128b8fe-d29b-40fc-b41f-b073f79543fc)

Using a one `query_string` (phrase):
This approach uses a `query_string` phrase type.

The main issue here is that it only returns results for documents that contain the search terms in the provided order, and documents like those shown before (Loretta Lynch) are not returned in the results.
![Screenshot 2023-06-06 at 0 03 31](https://github.com/freelawproject/courtlistener/assets/486004/3e687f0d-e326-496e-8c66-d6690254210f)

Also highlighting is not working as expected in synonyms.
![Screenshot 2023-06-06 at 0 00 55](https://github.com/freelawproject/courtlistener/assets/486004/11fb2055-3164-47c0-8bd9-b4ee809cb0b2)

So consider this information, what do you think is the best approach? Using the two-query approach (limiting the fields in both queries) or moving to a single `query_string`?


Here are the assets I use to perform this benchmark so it can be reproduced.

These files should be placed in the same directory and converted to .json:
[track.txt](https://github.com/freelawproject/courtlistener/files/11667101/track.txt)
[mappings.txt](https://github.com/freelawproject/courtlistener/files/11667104/mappings.txt)
[search_1.txt](https://github.com/freelawproject/courtlistener/files/11667111/search_1.txt)
[search_2.txt](https://github.com/freelawproject/courtlistener/files/11667112/search_2.txt)
[search_3.txt](https://github.com/freelawproject/courtlistener/files/11667113/search_3.txt)
[documents.txt](https://github.com/freelawproject/courtlistener/files/11667158/documents.txt)

I create this script to extract data from my local index and write it into documents.json:
[extract_docs.py.txt](https://github.com/freelawproject/courtlistener/files/11667134/extract_docs.py.txt)

To run the benchmark:

esrally should be installed:
`pip3 install esrally`

And then run:
`esrally race --distribution-version=7.16.3 --track-path=/path/with/project_json_files`

Reply:
Per our conversation, this is a quick analysis of the three options:

  | Q1: Two q strings, one fielded | Q2: Two q strings, both fielded | Q3: 1 q string
-- | -- | -- | --
Performance | Good | Great | Best
Highlighting | Good | Good | Misses fields
Relevancy | Good | Good | Worse
**Decision** | Good | Best | Not great

We won't change this here, but we'll get it in the next PR about real time OA alerts.